libraries/SDFS/src/SDFS.h

```diff
@@ -0,0 +1,404 @@
+        info.maxPathLength = 255; // TODO ?
+        info.totalBytes =_fs.vol()->volumeBlockCount() * 512;
+        info.usedBytes = info.totalBytes - (_fs.vol()->freeClusterCount() * _fs.vol()->blocksPerCluster() * 512);
+        return true;
+    }
+
+    bool remove(const char* path) override {
+        return _mounted ? _fs.remove(path) : false;
+    }
+
+    bool mkdir(const char* path) override {
+        return _mounted ? _fs.mkdir(path) : false;
+    }
+
+    bool rmdir(const char* path) override {
+        return _mounted ?_fs.rmdir(path) : false;
+    }
+
+    bool setConfig(const FSConfig *cfg) override
+    {
+        if ((cfg->_type != SDFSConfig::fsid::FSId) || _mounted) {
+            return false;
+        }
+	_cfg = *static_cast<const SDFSConfig *>(cfg);
+        return true;
+    }
+
+    bool begin() override {
+        if (_mounted) {
+            end();
+        }
+        _mounted = _fs.begin(_cfg._csPin, _cfg._spiSettings);
+        if (!_mounted && _cfg._autoFormat) {
+            format();
+            _mounted = _fs.begin(_cfg._csPin, _cfg._spiSettings);
+        }
+        return _mounted;
+    }
+
+    void end() override {
+        _mounted = false;
+        // TODO
+    }
+
+    bool format() override;
+
+protected:
+    friend class SDFileImpl;
+    friend class SDFSDirImpl;
+
+    sdfat::SdFat* getFs()
+    {
+        return &_fs;
+    }
+
+    static uint8_t _getFlags(OpenMode openMode, AccessMode accessMode) {
+        uint8_t mode = 0;
+        if (openMode & OM_CREATE) {
+            mode |= sdfat::O_CREAT;
+        }
+        if (openMode & OM_APPEND) {
+            mode |= sdfat::O_AT_END;
+        }
+        if (openMode & OM_TRUNCATE) {
+            mode |= sdfat::O_TRUNC;
+        }
+        if (accessMode & AM_READ) {
+            mode |= sdfat::O_READ;
+        }
+        if (accessMode & AM_WRITE) {
+            mode |= sdfat::O_WRITE;
+        }
+        return mode;
+    }
+
+    sdfat::SdFat _fs;
+    SDFSConfig   _cfg;
+    bool         _mounted;
+};
+
+
+class SDFSFileImpl : public FileImpl
+{
+public:
+    SDFSFileImpl(SDFSImpl *fs, std::shared_ptr<sdfat::File> fd, const char *name)
+        : _fs(fs), _fd(fd), _opened(true)
+    {
+        _name = std::shared_ptr<char>(new char[strlen(name) + 1], std::default_delete<char[]>());
+        strcpy(_name.get(), name);
+    }
+
+    ~SDFSFileImpl() override
+    {
+        flush();
+        close();
+    }
+
+    size_t write(const uint8_t *buf, size_t size) override
+    {
+        return _opened ? _fd->write(buf, size) : -1;
+    }
+
+    size_t read(uint8_t* buf, size_t size) override
+    {
+        return _opened ? _fd->read(buf, size) : -1;
+    }
+
+    void flush() override
+    {
+        if (_opened) {
+            _fd->flush();
+            _fd->sync();
+        }
+    }
+
+    bool seek(uint32_t pos, SeekMode mode) override
+    {
+        if (!_opened) {
+            return false;
+        }
+        switch (mode) {
```

Comment:
It's handled now with the `return false` catch-all at the end of the method.  I think it's more explicit there than moved into a case (since the fcn has more than just the case in it).  Thoughts?

Reply:
Ok, I should explain better. The default case in a switch should be implemented in all cases. For those cases where it doesn't make sense, I.e. should never happen, it usually contains an assert. It's a code robustness thing, it guards against a case that, well, should never happen. If it does happen, it means something is very wrong code wise, e.g. corruption, in which case a silent return false is not going to give any clues.

Reply:
Added in some debug messages on bad calls in general and a default handler in this specific case.

Reply:
You don't think an assert is valid here?

Reply:
OK, you win.  Just added with a condition that's guaranteed to fail should it be hit.  Still returning false afterwards for release builds so the file open will fail instead of be in an indeterminate state.